// version: 0.0.11
// date: 19/07/2020
syntax = "proto3";
package ign;
//-----------------------------------------------
// Main message types
//-----------------------------------------------
enum EventType {
  UNSPECIFIED        = 0;
  // ---Bridge [1~127]---
  // Device heartbeat
  HEARTBEAT          = 1;
  // Event contains activity log
  LOCK_LOG           = 2;
  // Get user account information
  GET_USER_INFO      = 3;
  // Updates server on connection test results of the selected locks
  LOCKS_TEST_UPDATE  = 4;
  // Adds a lock to bridge's list of connected lock
  ADD_LOCK           = 5;
  // Request for jobs
  GET_JOBS           = 6;
  // Update a job's status
  UPDATE_JOB         = 7;
  // Update battery (demo)
  DEMO_UPDATE_LOCK_BATTERY = 8;
  // Update lock status (demo)
  DEMO_UPDATE_LOCK_STATUS = 9;
  // Update activity logs (demo)
  DEMO_UPDATE_LOCK_ACTIVITIES = 10;
  // ---Server [128~255]---
  // Update user information on bridge
  UPDATE_USER_INFO   = 128;
  // Contains a job
  LOCK_JOB           = 129;
  // New job notification
  NEW_JOB_NOTIFY     = 130;
}
message MsgInfo {
  uint32          msg_id      = 1;
  uint32          time        = 2;
  EventType       event_type  = 3;
  BridgeEventData bridge_data = 4;
  ServerEventData server_data = 5;
}
//-----------------------------------------------
// Bridge event types
//-----------------------------------------------
message BridgeEventData {
  BridgeProfile profile = 1;
  // List of lock bt id
  repeated string bt_ids = 2;
  // List of activity logs
  repeated LockLog lock_log = 3;
  // List of lock id that failed the connection test
  LocksTestUpdate locks_test_update = 4;
  // Demo events
  string demo_lockId = 16;
  DemoUpdateLockBattery demo_update_lock_battery = 17;
  DemoUpdateLockStatus demo_update_lock_status = 18;
  DemoUpdateLockActivities demo_update_lock_activities = 19;
}
enum OSType {
  UKNOWN = 0;
  LINUX  = 1;
  RTOS   = 2;
}
message BridgeProfile {
  OSType os_info     = 1;
  bytes  bt_id       = 2;
  bytes  mac_addr    = 3;
  bytes  local_ip    = 4;
  bytes  public_ip   = 5;
  // sys_statics:
  //   byte[0] loader
  //   byte[1] cpu_rate
  //   byte[2] mem_rate
  //   byte[3] disk_rate
  bytes  sys_statics = 6;
  bytes  wifi_ssid   = 7;
  int32  wifi_signal = 8;
  uint32 inited_time = 9;
  bytes  name        = 10;
}
message LockLog {
  // Bt id of lock that this log belongs to
  string bt_id = 1;
  // Contains (admin key) encrypted activity logs
  bytes  log_data = 2;
}
message LocksTestUpdate {
  // Bt id of lock that failed connection test
  repeated string bt_id = 1;
}
message DemoUpdateLockBattery {
  uint32 battery = 1;
}
message DemoUpdateLockStatus {
  uint32 status = 1;
}
message DemoUpdateLockActivities {
  bytes log = 1;
}
//-----------------------------------------------
// Server event types
//-----------------------------------------------
message LockJob {
  // Bt id of lock that this job is for
  string bt_id    = 1;
  // Contains (admin key) encrypted bluetooth commands
  bytes  lock_cmd = 2;
}
message Ekey {
  bytes guest_aes_key = 1;
  bytes guest_token   = 2;
  bytes password      = 3;
  // Random 31-bit integer
  int32 keyId         = 4;
}
message LockEntry {
  // Lock bluetooth ID
  string bt_id = 1;
  // 'ekey' (with permissions) used to communicate with lock
  Ekey   ekey  = 2;
}
message ServerEventData {
  // List of lock entries
  repeated LockEntry lock_entries = 1;
  // Job belonging to a lock
  LockJob            job          = 2;
  DemoLockJob        demo_job     = 3;
}
// DEMO USE ONLY //
enum DemoLockCommand {
  UNSPECIFIED_DLC = 0;
  UNLOCK          = 1;
  LOCK            = 2;
  CREATE_PIN      = 3;
  DELETE_PIN      = 4;
  GET_BATTERY     = 5;
  GET_LOCK_STATUS = 6;
  GET_LOGS        = 7;
}
// DEMO USE ONLY //
message DemoLockJob {
  string          bt_id    = 1;
  DemoLockCommand op_cmd   = 2;
  // 4-8 digit pin (ascii)
  bytes           pin      = 3;
}
